vault/expiration.go

```diff
@@ -125,46 +128,111 @@ func (m *ExpirationManager) Restore() error {
 	if err != nil {
 		return fmt.Errorf("failed to scan for leases: %v", err)
 	}
-
 	m.logger.Debug("expiration: leases collected", "num_existing", len(existing))
 
-	// Restore each key
-	for i, leaseID := range existing {
-		if i%500 == 0 {
-			m.logger.Trace("expiration: leases loading", "progress", i)
+	// Make the channels used for the worker pool
+	broker := make(chan string)
+	quit := make(chan bool)
+	// Buffer these channels to prevent deadlocks
+	errs := make(chan error, len(existing))
+	result := make(chan *leaseEntry, len(existing))
+
+	// Use a wait group
+	wg := &sync.WaitGroup{}
+
+	// Create 64 workers to distribute work to
+	for i := 0; i < RestoreWorkerCount; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			for {
+				select {
+				case leaseID, ok := <-broker:
+					// broker has been closed, we are done
+					if !ok {
+						return
+					}
+
+					le, err := m.loadEntry(leaseID)
+					if err != nil {
+						errs <- err
+						continue
+					}
+
+					// Write results out to the result channel
+					result <- le
+
+				// quit early
+				case <-quit:
+					return
+				}
+			}
+		}()
+	}
+
+	// Distribute the collected keys to the workers in a go routine
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for i, leaseID := range existing {
+			if i%500 == 0 {
+				m.logger.Trace("expiration: leases loading", "progress", i)
+			}
+
+			select {
+			case <-quit:
+				return
+
+			default:
+				broker <- leaseID
+			}
 		}
-		// Load the entry
-		le, err := m.loadEntry(leaseID)
-		if err != nil {
+
+		// Close the broker, causing worker routines to exit
+		close(broker)
```

Comment:
Because broker is not buffered, by the time the execution get here all the leases should have been loaded.